sql: support read-only transactions

jordanlewis opened the pull request:
Add support for:

- SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY/READ WRITE;
- SET TRANSACTION READ ONLY/READ WRITE;
- SET default_transaction_read_only;
- SET transaction_read_only;

Fixes #16033.

Release note: Add support for read-only transactions via Postgres
compatible syntax.

jordanlewis committed:
sql: support read-only transactions

Add support for:

- SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY/READ WRITE;
- SET TRANSACTION READ ONLY/READ WRITE;
- SET default_transaction_read_only;
- SET transaction_read_only;

Release note: Add support for read-only transactions via Postgres
compatible syntax.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20547)
<!-- Reviewable:end -->

justinj commented on pkg/sql/sem/tree/stmt.go:
```diff
@@ -101,6 +101,18 @@ func CanModifySchema(stmt Statement) bool {
+	}
+	return false
+}
+
```
Can our `[...]` syntax evade this check? I think we might need to recurse? Also
should sequence functions like `nextval` etc. count as writes within this context?
cc @vilterp

jordanlewis replied:
We call this in `newPlan` which does get recursively called for all statements, so I *think* we're set here.

justinj replied:
Ahh true, I missed that 👍

vilterp replied:
yeah, `SELECT nextval('foo')` can cause a write, but those functions also spawn their own transaction, which immediately completes (you can't roll it back). So maybe the containing transaction would still sort of be read only?

justinj commented on pkg/sql/session.go:
```diff
@@ -999,6 +1002,9 @@ type txnState struct {
 
+	// THe transaction's read only state.
+	readOnly bool
+
```
TAke a look y'all

jordanlewis replied:
Done.

justinj reviewed (commented):
Random thought: can we automatically demote read-only transactions to SNAPSHOT to some benefit?
<!-- review by re -->

knz commented:
Nice move compatibility. I can see uses for testing too.

Some more work needed though:

1) the BEGIN syntax also supports setting the read-only bit, can you please check for consistency with SET

2) you forgot sequences (`nextval`)

3) you forgot `Split`

4) you forgot RESTORE (doesn't do RowsAffected but most certainly writes)

5) you may have forgotten COPY

---

Reviewed 13 of 14 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sem/tree/stmt.go, line 106 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20547#-L-lVoByJmWucnyaooj8:-L-lVoByJmWucnyaooj9:b-aujx8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b944af27b5c38479f1506fc36d2143986a55f0b/pkg/sql/sem/tree/stmt.go#L106)):*
> ```Go
> // CanWriteData returns true if the statement can modify data.
> func CanWriteData(stmt Statement) bool {
> 	if stmt.StatementType() == RowsAffected {
> ```

I'm not fond of this condition - what says that RowsAffected statements can write? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20547#-:-L-lWUQxjcdaxPTCJOTs:b-1w0k5l)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Instead of what you're doing, I think enforcement of read only should be done at the txn level (`client.Txn`). That guy already knows if it's doing reads or writes.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20547#-:-L-lp7pEevlPpKJvtlyX:b-fluavj)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
@andreimatei I don't think that will work - the semantics are that statements that *could* write are banned. If a statement that could write but does not actually write succeeds in a read only transaction, that would be a violation of the contract.

knz commented:
Also the same `client.Txn` is used to write other things - table descriptor lease info, events, etc. These writes should go through.

andreimatei commented:
@jordanlewis 
If a statement doesn't write, it's read only in my book.
You're saying that it can be statically determined whether a statement should be rejected or not. That means that a user should never send such a statement (they can easily do the determination themselves). So the difference in semantics would only be observed by users who send such statements anyway. For those, I don't really see the problem in accepting the statement. As far as I'm concerned, no harm - no foul.

@knz 
If transactions really need to write stuff, then I'll give it to you. But do they really?
You don't need to take a lease if you don't want to; leases are only about caching for future timestamps - that's optional.
Events... what events exactly? Aren't those for schema changes?


Here's where I'm coming from: I think a use for such read-only txns should be to ensure that a transaction truly writes nothing - i.e. it cannot mess up the database, or get stuck in ways that writes can get stuck when shit's fucked (this about reading things when the database cannot write anything). To support this, a txn truly has to write nothing, beyond just not writing "user data". So not writing new leases would be a good thing, not a bad one.
I realize that such a feature doesn't _need_ to be the same as the one implemented here for the purposes of compatibility with whatever, but I think we can shoot all the rabbits with this occasion. Also, implementing this feature at a lower level ensures safety for someone who actually cares about read-only txns - e.g. if you're using this feature because you're handing a txn to untrusted code (otherwise, why exactly would you use this feature?), then we'd be more confident that we can enforce it if we enforce it at a lower level (like, there's already allegations in this review that stuff was missed).

What do y'all think?

andreimatei commented:
Btw, leases don't technically fall into this discussion at all because they're not written within the transaction that's trying to get a table descriptor; they use separate transactions.
But my point stands that readonly txns should not trigger these lease writes (or at least not block for them to complete).

jordanlewis commented:
I'm not okay with diverging from the semantics of `READ ONLY`, which are defined in the SQL standard:

```
         The operations of insert, update, and delete are
         permitted for updatable tables, subject to constraining Access
         Rules. The operations of insert, update, and delete are not allowed
         for read-only tables.
```

This requires the static checking we're discussing here.

I would be okay with discussing a more CockroachDB-specific approach that uses the Txn, but I believe that's a separate task. In the short term, we need to implement the standard-mandated semantics.

jordanlewis commented:
Sorry, that clause is the wrong clause to reference. That's about read-only tables, not read-only transactions.

The correct clause is repeated across all of the write statements, and looks like this:

```
            1) If the access mode of the current SQL-transaction is read-only
            and T is not a temporary table, then an exception condition is
            raised: invalid transaction state.
```

jordanlewis commented:
@knz

> 1. the BEGIN syntax also supports setting the read-only bit, can you please check for consistency with SET.

This was tested in the original PR.

> 2. you forgot sequences (nextval)

Done.

> 3. you forgot Split

Done as well as Truncate, TestingRelocate and Scatter.

> 4. you forgot RESTORE (doesn't do RowsAffected but most certainly writes)

Done as well as IMPORT.

> 5. you may have forgotten COPY

Done.

---

Review status: 10 of 15 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sem/tree/stmt.go, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20547#-L-jexYw1gHOpINWXJ48-r1-115:-L-n0xJbqoMMSh6xfl7D:b-1dlpkf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9d60ad349ddfc10ffbe65d90b8b5d21a30eb72f/pkg/sql/sem/tree/stmt.go#L115)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

yeah, `SELECT nextval('foo')` can cause a write, but those functions also spawn their own transaction, which immediately completes (you can't roll it back). So maybe the containing transaction would still sort of be read only?
</blockquote></details>

Fixed nextVal.

---

*[pkg/sql/sem/tree/stmt.go, line 106 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20547#-L-lVoByJmWucnyaooj8:-L-n0wOqU7-xIPjvE6dQ:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b944af27b5c38479f1506fc36d2143986a55f0b/pkg/sql/sem/tree/stmt.go#L106)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I'm not fond of this condition - what says that RowsAffected statements can write? 
</blockquote></details>

Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20547#-:-L-n14fbsc3qJOHlrG5w:bk6rbdm)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r2, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20547#-:-L-n2Xneu4yLT1RVWRI8:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
@justinj regarding SNAPSHOT, I have no idea, but even if that were theoretically true I believe our SNAPSHOT transactions are sufficiently untested that I would be afraid to do such a thing.

andreimatei commented:
I don't care if we reject all INSERTS/UPDATES/DELETES when in this read-only mode; that I don't have a problem with. We can do static checking if you want. I'm not sure what the best way to do it is, though. How long until we have something besides `Insert, Delete, Update` that we want to reject? Or how long until our handling of subqueries improves and we no longer call this `newPlan` method for every subquery? Or how long until we have Insert,etc that we _don't_ want to reject? Like, from what I'm reading, both MySQL and Postgres allow modification of temp tables. I'm thinking that we should make this a property of the plan nodes, and compute it recursively.

About the rest the truth is that I don't care too much so I'm no longer insisting on anything, but still, let's philosophize:

I'd be interested in stepping back and answering what is the definition of read-only that we want? This should be informed by those temp tables, and also by Postgres' guarantee that read-only serializable transactions are _guaranteed_ to not encounter serializability violations. Do we like this? If we do (and I do!), then I think that we have to enforce that really nothing gets written at a lower level. And if we do that, then the next question is if this should extend to writes that are not technically done within the transaction, but morally are caused by it. Like the `SELECT nextval()` thing; you didn't yet take a position on that and I'm curious. And also the lease acquisitions.

I want to preempt accusations of talking myself to death by half-jokingly suggesting that there is a way to provide compatibility with ORMs for this feature - support the syntax and do nothing with it. The only users that'd observe the difference are ones that use the feature... dare I say... wrongly.
I, for one, would not mind us doing that because I have to say that I don't really like/understand this feature, as it is presented in SQL. It's clearly not a security-oriented feature - cause I can't imagine how you'd actually use it for that. The reason why I *think* this syntax made its way into the standard is to provide a hint to the database about what's coming in the hope that it can do optimizations. We don't have such optimizations to make at the moment (although the UCSB people were saying that their "dynamic timestamp" thing that they've been working on depends on knowing that transactions are r/o ahead of time, but I don't know if their thing is valid or not), so at the moment I don't particularly feel a need to enforce it.

OTOH, as I was saying, what I personally do think I like is the idea of r/o txn defined at a low level - knowing that a txn will leave the database just as it found it and that it can operate on a r/o `Engine` - for example, on a mounted backup image (which @danhhz said he'll do and I thought it'd be the coolest thing ever). And if you define it like that, then I think we should also implement it at a lower level.

jordanlewis commented:
> I don't care if we reject all INSERTS/UPDATES/DELETES when in this read-only mode; that I don't have a problem with. We can do static checking if you want. I'm not sure what the best way to do it is, though. How long until we have something besides Insert, Delete, Update that we want to reject? Or how long until our handling of subqueries improves and we no longer call this newPlan method for every subquery? Or how long until we have Insert,etc that we don't want to reject? Like, from what I'm reading, both MySQL and Postgres allow modification of temp tables. I'm thinking that we should make this a property of the plan nodes, and compute it recursively.

This is a good point. I stuck with this approach because it's a lot simpler - the handling of determining whether a read-only violation is occurring is centralized to a single place. But perhaps it would be better to scatter checks in the `newPlan` implementations of all of the potentially-writing plan nodes. I think though that this refactor is simple enough that it's not too much of a burden to push it off until we need it for temporary tables.

For subqueries, I'm not sure what the problem is - unless I'm forgetting something, subqueries are syntax-enforced to be read-only (`SELECT`). Our `[]` syntax of course considerably complicates this story, but assuming we're leaving that in as a pure convenience (as I would like) then we should always end up invoking `newPlan` on it.

> I'd be interested in stepping back and answering what is the definition of read-only that we want? This should be informed by those temp tables, and also by Postgres' guarantee that read-only serializable transactions are guaranteed to not encounter serializability violations. Do we like this? If we do (and I do!), then I think that we have to enforce that really nothing gets written at a lower level. And if we do that, then the next question is if this should extend to writes that are not technically done within the transaction, but morally are caused by it. Like the SELECT nextval() thing; you didn't yet take a position on that and I'm curious. And also the lease acquisitions.

I don't understand enough about lease writes to comment on that. My intuition is that if Cockroach needs to make some writes to do reads, then those writes should be permitted, but again I'm not familiar enough with the details here. `nextval()` should be prevented by this read-only mode.

> I want to preempt accusations of talking myself to death by half-jokingly suggesting that there is a way to provide compatibility with ORMs for this feature - support the syntax and do nothing with it. 

This is a non-starter in my opinion. We are constantly being burnt by compatibility issues, and introducing new ones will only make our lives more difficult down the line. It may or may not be useful from your perspective, but many ORMs and tools use this feature in unknown ways. Perhaps they do use it as a safety net to prevent developers using those tools from doing incorrect things. We can't know - and therefore in my opinion we can't take this suggestion.


> OTOH, as I was saying, what I personally do think I like is the idea of r/o txn defined at a low level - knowing that a txn will leave the database just as it found it and that it can operate on a r/o Engine - for example, on a mounted backup image (which @danhhz said he'll do and I thought it'd be the coolest thing ever). And if you define it like that, then I think we should also implement it at a lower level.

Agree that this is a good idea at some point. But I don't see any need to bundle these two changes together - I think we would benefit from collecting some more requirements on this r/o Engine and other related issues before diving into an r/o transaction implementation.

jordanlewis commented:
TFTRs! Andrei and I continued this conversation offline.

The remaining question for this PR was regarding preventing builtins that write from executing under read-only mode - Andrei suggested making this a builtin attribute instead of having one-off checks.

After some consideration, I've decided to keep the one-off checks. I can't think of any builtin functions (or ideas for them) besides `nextVal` and `setVal` that behave like this, so in my opinion the overhead of adding this attribute to all builtins is unjustified.
